domain/machine/state/reboot.go

```diff
@@ -0,0 +1,197 @@
+// Copyright 2024 Canonical Ltd.
+// Licensed under the AGPLv3, see LICENCE file for details.
+
+package state
+
+import (
+	"context"
+
+	"github.com/canonical/sqlair"
+	"github.com/juju/errors"
+
+	machineerrors "github.com/juju/juju/domain/machine/errors"
+	"github.com/juju/juju/internal/database"
+)
+
+// RebootAction defines the action a machine should
+// take when a hook needs to reboot
+type RebootAction string
+
+const (
+	// ShouldDoNothing instructs a machine agent that no action
+	// is required on its part
+	ShouldDoNothing RebootAction = "noop"
+	// ShouldReboot instructs a machine to reboot
+	// this happens when a hook running on a machine, requests
+	// a reboot
+	ShouldReboot RebootAction = "reboot"
+	// ShouldShutdown instructs a machine to shut down. This usually
+	// happens when running inside a container, and a hook on the parent
+	// machine requests a reboot
+	ShouldShutdown RebootAction = "shutdown"
+)
+
+// RequireMachineReboot sets the machine referenced by its UUID as requiring a reboot.
+//
+// Reboot requests are handled through the "machine_requires_reboot" table which contains only
+// machine UUID for which a reboot has been requested.
+// This function is idempotent.
+func (st *State) RequireMachineReboot(ctx context.Context, uuid string) error {
+	db, err := st.DB()
+	if err != nil {
+		return errors.Trace(err)
+	}
+
+	setRebootFlag := `INSERT INTO machine_requires_reboot (machine_uuid) VALUES ($machineUUID.uuid)`
+	setRebootFlagStmt, err := sqlair.Prepare(setRebootFlag, machineUUID{})
+	if err != nil {
+		return errors.Trace(err)
+	}
+	err = db.Txn(ctx, func(ctx context.Context, tx *sqlair.TX) error {
+		return tx.Query(ctx, setRebootFlagStmt, machineUUID{uuid}).Run()
+	})
+	if database.IsErrConstraintPrimaryKey(err) {
+		// if the same uuid is added twice, do nothing (idempotency)
+		return nil
+	}
```

Comment:
With https://github.com/juju/juju/pull/17757 coming down the line, it will be required to handle the error at the source. So, move this into `db.Txn` function.